src/backend/pipeline/trigger/trigger.c

```diff
@@ -0,0 +1,1181 @@
+
+	heap_close(rel, AccessShareLock);
+	ReleaseSysCache(tup);
+}
+
+/*
+ * synchronize
+ *
+ * Create a sync batch, and scan the pipeline_query system catalog for CQs.
+ * Call do_decode_change for each one.
+ *
+ * This will have the effect of adding a bunch of empty changelists for each rel
+ * to a batch.
+ *
+ * process_batches will subsequently pick up this batch and
+ * run diff_triggers on it etc
+ */
+static void
+do_synchronize(TriggerProcessState *state)
+{
+	HeapTuple tup;
+	XactBatch *batch = start_new_batch(state, "sync", GetTopTransactionId(),
+			GetCurrentTimestamp());
+
+	RangeVar *rv = makeRangeVar(NULL, "pipeline_query", 0);
+	Relation prel = heap_openrv(rv, AccessShareLock);
+
+	HeapScanDesc scan = heap_beginscan(prel, GetTransactionSnapshot(), 0, NULL);
+
+	while ((tup = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	{
+		Form_pipeline_query row = (Form_pipeline_query) GETSTRUCT(tup);
+		Oid relid = row->matrel;
+
+		Relation rel = heap_open(relid, AccessShareLock);
+
+		do_decode_change(state, batch, rel, TRIGGER_PROCESS_CHANGE_NOOP,
+				NULL, NULL);
+
+		heap_close(rel, AccessShareLock);
+	}
+
+	heap_endscan(scan);
+	heap_close(prel, NoLock);
+
+	batch->finished = true;
+}
+
+/*
+ * synchronize
+ *
+ * Create a sync batch inside a transaction.
+ * (This is done in response to a HUP).
+ *
+ * The main use of this is to ensure internal state is synchronized before
+ * any inserts or updates are seen.
+ */
+static void
+synchronize(TriggerProcessState *state)
+{
+	StartTransactionCommand();
+	do_synchronize(state);
+	CommitTransactionCommand();
+}
+
+/*
+ * do_decode_change
+ *
+ * Add a change (action, old_tup, new_tup) to a relation to a batch.
+ *
+ * Some basic book keeping of TriggerCacheEntry is performed here,
+ * and also some simple filtering.
+ *
+ * We ignore changes to non CV matrels, and drop any changes that occur before
+ * the batch processor has noticed triggers on a CV. entry->numtriggers is
+ * used to track that state.
+ */
+static void
+do_decode_change(TriggerProcessState *state,
+				 XactBatch *xact_batch,
+				 Relation rel,
+				 TriggerProcessChangeType action,
+				 HeapTuple old_tup,
+				 HeapTuple new_tup)
+{
+	ChangeList *cl;
+	TriggerCacheEntry *entry;
+	Oid	relid;
+	bool found;
+
+	relid = RelationGetRelid(rel);
+
+	entry = hash_search(state->trigger_cache, &relid,
+				HASH_ENTER, &found);
+
+	if (!found)
+	{
+		memset(entry, 0, sizeof(TriggerCacheEntry));
+		entry->matrelid = relid;
+		GetCVInfo(rel, entry, state->cache_cxt);
+	}
+
+	if (entry->cvrelid == InvalidOid || entry->is_adhoc)
+		return;
+
+	/* This next line is important, because we *do* want to pass empty
+	 * changelists to the batch processor so that it can notice new triggers
+	 * etc */
+
+	cl = get_changelist(state, xact_batch, relid);
+
+	if (action == TRIGGER_PROCESS_CHANGE_DELETE ||
+	 	action == TRIGGER_PROCESS_CHANGE_NOOP)
+		return;
+
+	if (!entry->numtriggers)
+		return;
+
+	// TRIGGER
```

Comment:
Yeah, if you see any // comments from me, they are almost always placeholders that have been erroneously leftover. I usually grep for them before committing to master.